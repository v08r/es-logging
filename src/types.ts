import type {
    AlertCode,
    Code,
    CriticalCode,
    DebugCode,
    EmergencyCode,
    ErrorCode,
    InfoCode,
    NoticeCode,
    WarningCode,
} from './codes';
import type { Severity } from './constants';
import type Entry from './Entry';

type AllowedCode<S extends Severity, C extends Code> = AllowedSeverity<C, SeverityFromCode<C>, S>;

type AllowedSeverity<
    C extends Code,
    S extends Severity,
    T extends Severity
> = S extends Severity.Emergency
    ? C
    : T extends Severity.Alert
    ? S extends Severity.Emergency | Severity.Alert
        ? C
        : never
    : T extends Severity.Critical
    ? S extends Severity.Emergency | Severity.Alert | Severity.Critical
        ? C
        : never
    : T extends Severity.Error
    ? S extends Severity.Emergency | Severity.Alert | Severity.Critical | Severity.Error
        ? C
        : never
    : T extends Severity.Warning
    ? S extends
          | Severity.Emergency
          | Severity.Alert
          | Severity.Critical
          | Severity.Error
          | Severity.Warning
        ? C
        : never
    : T extends Severity.Notice
    ? S extends
          | Severity.Emergency
          | Severity.Alert
          | Severity.Critical
          | Severity.Error
          | Severity.Warning
          | Severity.Notice
        ? C
        : never
    : T extends Severity.Info
    ? S extends
          | Severity.Emergency
          | Severity.Alert
          | Severity.Critical
          | Severity.Error
          | Severity.Warning
          | Severity.Notice
          | Severity.Info
        ? C
        : never
    : C;

interface Constructor<T = {}> {
    new (...args: any[]): T;
}

type EntryDefinitions<C extends Code> = Record<C, { message: string; payload: any }>;

type KnownEntry<
    PID extends string,
    PV extends number,
    C extends Code,
    ED extends Record<
        C,
        {
            message: string;
            payload: any;
        }
    >
> =
    | ProductEntry<PID, PV, C, ED>
    | ProductEntry<
          PID,
          PV,
          LoggerReservedEntryCode,
          {
              300998: {
                  message: `logger level has been decreased from ${Exclude<
                      SeverityLabel,
                      'Debug'
                  >} to ${Exclude<SeverityLabel, 'Off'>}`;
                  payload: { from: Exclude<Severity, Severity.Debug> | null; to: Severity };
              };
              300999: {
                  message: 'logger has been attached';
                  payload: null;
              };
              400998: {
                  message: `logger level has been increased from ${Exclude<
                      SeverityLabel,
                      'Off'
                  >} to ${Exclude<SeverityLabel, 'Debug'>}`;
                  payload: { from: Severity; to: Exclude<Severity, Severity.Debug> | null };
              };
              400999: {
                  message: 'logger has been detached';
                  payload: null;
              };
          }
      >;

interface Logger<E extends Entry<number, string, number, Code, string, unknown>> {
    log(entry: E): void;
}

type LoggerReservedEntryCode = 300998 | 300999 | 400998 | 400999;

interface LoggerWithSeverity<C extends Code, ED extends EntryDefinitions<C>, S extends Severity> {
    <EC extends AllowedCode<S, C>>(
        code: EC,
        message: ED[EC]['message'],
        payload: ED[EC]['payload']
    ): void;
}

type MaybeReadonly<T> = T | Readonly<T>;

type ProductEntry<
    PID extends string,
    PV extends number,
    C extends Code,
    ED extends Record<
        C,
        {
            message: string;
            payload: any;
        }
    >
> = Value<{
    [k in Extract<keyof ED, Code>]: Entry<number, PID, PV, C, ED[k]['message'], ED[k]['payload']>;
}>;

type SeverityFromCode<C extends Code> = C extends DebugCode
    ? Severity.Debug
    : C extends InfoCode
    ? Severity.Info
    : C extends NoticeCode
    ? Severity.Notice
    : C extends WarningCode
    ? Severity.Warning
    : C extends ErrorCode
    ? Severity.Error
    : C extends CriticalCode
    ? Severity.Critical
    : C extends AlertCode
    ? Severity.Alert
    : C extends EmergencyCode
    ? Severity.Emergency
    : Severity;

type SeverityLabel = keyof typeof Severity | 'Off';

type Value<T> = T[keyof T];

export type {
    AllowedCode,
    Code,
    Constructor,
    EntryDefinitions,
    KnownEntry,
    Logger,
    LoggerReservedEntryCode,
    LoggerWithSeverity,
    MaybeReadonly,
    SeverityFromCode,
    SeverityLabel,
};
