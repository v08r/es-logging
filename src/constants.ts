/** Severity of log entries. */
enum Severity {
    /** Debug or trace information. */
    Debug = 100,
    /** Routine information, such as ongoing status or performance. */
    Info = 200,
    /**
     * Normal but significant events, such as start up, shut down, or
     * a configuration change.
     */
    Notice = 300,
    /** Warning events might cause problems. */
    Warning = 400,
    /** Error events are likely to cause problems. */
    Error = 500,
    /** Critical events cause more severe problems or outages. */
    Critical = 600,
    /** A person must take an action immediately. */
    Alert = 700,
    /** One or more systems are unusable.  */
    Emergency = 800,
}

export { Severity };
