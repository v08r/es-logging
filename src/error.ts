/** This error happens when an invalid timestamp is supplied. */
class InvalidTimestampError<T extends number | string | Date> extends Error {
    name = 'InvalidTimestampError';
    message!: `invalid timestamp: ${string}`;

    constructor(readonly timestamp: T) {
        super(`invalid timestamp ${timestamp}`);
    }
}

export { InvalidTimestampError };
