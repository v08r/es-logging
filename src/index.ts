export { Severity } from './constants';
export { default as Entry } from './Entry';
export { default as LogAdapter } from './LogAdapter';
export type { Code } from './types';
