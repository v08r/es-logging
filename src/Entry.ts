import type { Code } from './codes';
import { InvalidTimestampError } from './error';
import type { MaybeReadonly, SeverityFromCode } from './types';

const normalizeTimestamp = (ts?: number | string | Date | null | undefined) => {
    if (ts == null) return null;

    let lts = ts;
    if (typeof lts === 'string') lts = new Date(lts);
    if (typeof lts === 'object') lts = lts.getTime();

    if (isNaN(lts)) throw new InvalidTimestampError(ts);

    return lts;
};

/** Log entry. */
class Entry<
    TD extends number,
    PID extends string,
    PV extends number,
    C extends Code,
    M extends string,
    P
> {
    /**
     * [JSON reviver](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#the_reviver_parameter)
     * factory for `Entry`.
     *
     * @param keyFilter function to identify properties to be "revived". Pass
     *                  `(key: string) => key === ""` to parse the JSON root.
     */
    static reviver(keyFilter: (key: string) => boolean) {
        return (key: string, value: any) => {
            return keyFilter(key) &&
                Array.isArray(value) &&
                value.length === 8 &&
                (typeof value[0] === 'number' || value[0] === null) &&
                typeof value[1] === 'number' &&
                typeof value[2] === 'string' &&
                typeof value[3] === 'string' &&
                typeof value[4] === 'number' &&
                typeof value[5] === 'number' &&
                typeof value[6] === 'string'
                ? new Entry(value as any)
                : value;
        };
    }

    /** Log entry code. */
    readonly code: C;
    /**
     * Logger ID.
     *
     * NOTE: it is expected that a logger would mutate this property
     *       before logging the entry.
     */
    loggerID: string;
    /** Log entry message. */
    readonly message: M;
    /** Log entry payload. */
    readonly payload: P;
    /** ID of the software product creating the log entry. */
    readonly productID: PID;
    /** Version of the software product creating the log entry. */
    readonly productVersion: PV;
    /**
     * Milliseconds since [time origin](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)
     * when the log entry was created.
     */
    readonly timedelta: TD;
    /**
     * Milliseconds since the [ECMAScript epoch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#the_ecmascript_epoch_and_timestamps).
     * May be `null`.
     *
     * NOTE: it is expected that a logger would mutate this property
     *       before logging the entry.
     */
    readonly timestamp: number | null;

    /**
     * @param code log entry code.
     * @param loggerID logger ID.
     * @param message log entry message.
     * @param payload log entry payload.
     * @param productID ID of the software product creating the log entry.
     * @param productVersion Version of the software product
     *                       creating the log entry.
     * @param timedelta Milliseconds since [time origin](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin)
     *                  when the log entry was created.
     * @param timestamp Timestamp as milliseconds since the [ECMAScript epoch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#the_ecmascript_epoch_and_timestamps),
     *                  JSON representation for a `Date`, `Date` or `null`
     */
    constructor(
        code: C,
        loggerID: string,
        message: M,
        payload: P,
        productID: PID,
        productVersion: PV,
        timedelta: TD,
        timestamp: number | string | Date | null
    );
    /**
     * @param data Array including timestamp, timedelta, logger ID, product ID,
     *             product version, code, message and payload
     *             (in the mentioned order).
     */
    constructor(data: MaybeReadonly<[number | string | Date | null, TD, string, PID, PV, C, M, P]>);
    /**
     * @param json JSON representation for an `Entry`.
     */
    constructor(json: string);
    constructor(
        jsonDataOrCode:
            | string
            | MaybeReadonly<[number | string | Date | null, TD, string, PID, PV, C, M, P]>
            | C,
        loggerID?: string,
        message?: M,
        payload?: P,
        productID?: PID,
        productVersion?: PV,
        timedelta?: TD,
        timestamp?: number | string | Date | null
    ) {
        let c: C;
        let lid: string;
        let m: M;
        let p: P;
        let pid: PID;
        let pv: PV;
        let td: TD;
        let ts: number | string | Date | null | undefined;

        if (typeof jsonDataOrCode === 'number') {
            c = jsonDataOrCode;
            lid = loggerID as string;
            m = message as M;
            p = payload as P;
            pid = productID as PID;
            pv = productVersion as PV;
            td = timedelta as TD;
            ts = timestamp;
        } else
            [ts, td, lid, pid, pv, c, m, p] =
                typeof jsonDataOrCode === 'string' ? JSON.parse(jsonDataOrCode) : jsonDataOrCode;

        this.code = c;
        this.loggerID = lid;
        this.message = m;
        this.payload = p;
        this.productID = pid;
        this.productVersion = pv;
        this.timedelta = td;
        this.timestamp = normalizeTimestamp(ts);
    }

    /** Severity for the log entry. */
    get severity() {
        return Math.floor(this.code / 1000) as SeverityFromCode<C>;
    }

    /** Timestamp for the log entry. */
    getTimeStamp() {
        const { timestamp } = this;
        return timestamp === null ? null : new Date(timestamp);
    }

    /** Sets the timestamp for the log entry. */
    setTimestamp(timestamp: number | string | Date) {
        (this as any).timestamp = normalizeTimestamp(timestamp);
        return this;
    }

    /** Converts the `Entry` into an array suitable for JSON serialization. */
    toJSON() {
        return [
            this.timestamp,
            this.timedelta,
            this.loggerID,
            this.productID,
            this.productVersion,
            this.code,
            this.message,
            this.payload,
        ] as [number | null, TD, string, PID, PV, C, M, P];
    }
}

export { Entry as default };
