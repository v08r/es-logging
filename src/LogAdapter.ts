import type {
    AlertCode,
    Code,
    CriticalCode,
    DebugCode,
    EmergencyCode,
    ErrorCode,
    InfoCode,
    NoticeCode,
    WarningCode,
} from './codes';
import { Severity } from './constants';
import Entry from './Entry';
import type {
    EntryDefinitions,
    KnownEntry,
    Logger,
    LoggerReservedEntryCode,
    LoggerWithSeverity,
} from './types';

type LogMethod<C extends Code, ED extends EntryDefinitions<C>> =
    | (<DC extends C>(code: DC, message: ED[DC]['message'], payload: ED[DC]['payload']) => void)
    | undefined;

const severityValues = Object.values(Severity).reduce((sm, severity) => {
    if (typeof severity === 'number') sm.push([Severity[severity].toLowerCase(), severity]);
    return sm;
}, [] as Array<[string, Severity]>);

const setLog = (logAdapter: any, log: any, threshold: Severity | null) => {
    for (const [key, severity] of severityValues)
        logAdapter[key] = threshold && severity >= threshold ? log : undefined; // eslint-disable-line no-invalid-this
};

const logLevelChange = (
    logAdapter: any,
    previous: Severity | null,
    current: Severity | null,
    increase?: boolean
) => {
    logAdapter[increase ? 'warning' : 'notice']?.(
        increase ? 400998 : 300998,
        `logger level has been ${increase ? 'in' : 'de'}creased from ${
            Severity[previous as Severity] || 'Off'
        } to ${Severity[current as Severity] || 'Off'}`,
        null
    );
};

/** Adapter for loggers. */
class LogAdapter<
    PID extends string,
    PV extends number,
    C extends Exclude<Code, LoggerReservedEntryCode>,
    ED extends EntryDefinitions<C>
> {
    #logger?: Logger<KnownEntry<PID, PV, C, ED>>;
    #log?: LoggerWithSeverity<C, ED, Severity>;
    #severity: Severity | null = null;
    #productID: PID;
    #productVersion: PV;

    /**
     * @param productID ID of the software product associated with log entries.
     * @param productVersion Version of the software product associated
     *                       with log entries.
     */
    constructor(productID: PID, productVersion: PV) {
        this.#productID = productID;
        this.#productVersion = productVersion;
    }

    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Debug` and a logger is attached.
     */
    readonly debug?: LogMethod<Extract<C, DebugCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Info` or lower and a logger is attached.
     */
    readonly info?: LogMethod<Extract<C, InfoCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Notice` or lower and a logger is attached.
     */
    readonly notice?: LogMethod<Extract<C, NoticeCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Warning` or lower and a logger is attached.
     */
    readonly warning?: LogMethod<Extract<C, WarningCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Error` or lower and a logger is attached.
     */
    readonly error?: LogMethod<Extract<C, ErrorCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Critical` or lower and a logger is attached.
     */
    readonly critical?: LogMethod<Extract<C, CriticalCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Alert` or lower and a logger is attached.
     */
    readonly alert?: LogMethod<Extract<C, AlertCode>, ED>;
    /**
     * Log method that is only available if severity has been set to:
     * `Severity.Emergency` or lower and a logger is attached.
     */
    readonly emergency?: LogMethod<Extract<C, EmergencyCode>, ED>;

    /** Currently attached logger. */
    set logger(logger: Logger<KnownEntry<PID, PV, C, ED>> | undefined) {
        if (logger !== this.#logger) {
            this.warning?.(400999 as any, 'logger has been detached', null);

            this.#logger = logger;

            setLog(
                this,
                (this.#log =
                    logger &&
                    ((code: any, message: any, payload: any) => {
                        const timedelta = performance.now();
                        logger.log(
                            new Entry(
                                code,
                                '',
                                message,
                                payload,
                                this.#productID,
                                this.#productVersion,
                                timedelta,
                                null
                            ) as any
                        );
                    })),
                this.#severity
            );

            this.notice?.(300999 as any, 'logger has been attached', null);
        }
    }
    get logger() {
        return this.#logger;
    }

    /**
     * Current severity level.
     *
     * NOTE:
     * - severity will always be `null` if no logger is attached.
     * - set to `null` to turn logging off.
     */
    set severity(severity: Severity | null) {
        const previousSeverity = this.#severity;
        if (previousSeverity !== severity) {
            const increase = (previousSeverity || 1000) < (severity || 1000);

            if (increase) logLevelChange(this, previousSeverity, severity, true);

            setLog(this, this.#log, (this.#severity = severity));

            if (!increase) logLevelChange(this, previousSeverity, severity);
        }
    }
    get severity() {
        return this.#logger ? this.#severity : null;
    }
}

export { LogAdapter as default };
