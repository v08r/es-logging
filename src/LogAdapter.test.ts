import { Severity } from '../src/constants';
import LogAdapter from '../src/LogAdapter';

const replaceTimedeltaFromCalls = (calls: Array<Array<any>>) => {
    for (const call of calls) for (const arg of call) arg.timedelta = 42;
};

describe('`LogAdapter`', () => {
    const productID = 'productID';
    const productVersion = 1;

    const logger1 = { log: jest.fn() };
    const logger2 = { log: jest.fn() };

    const la = new LogAdapter(productID, productVersion);

    test('creation', () => {
        for (const k of [
            'debug',
            'info',
            'notice',
            'warning',
            'error',
            'critical',
            'alert',
            'emergency',
            'logger',
        ] as const)
            expect(la[k]).toBe(undefined);
        expect(la.severity).toBe(null);
    });

    test('lifecycle', () => {
        la.severity = Severity.Error;
        expect(la.severity).toBe(null);

        la.logger = logger1;
        expect(la.logger).toBe(logger1);
        expect(la.severity).toBe(Severity.Error);
        for (const k of ['debug', 'info', 'notice', 'warning'] as const)
            expect(la[k]).toBe(undefined);
        for (const k of ['error', 'critical', 'alert', 'emergency'] as const)
            expect(typeof la[k]).toBe('function');

        la.severity = Severity.Notice;
        expect(la.severity).toBe(Severity.Notice);
        for (const k of ['debug', 'info'] as const) expect(la[k]).toBe(undefined);
        for (const k of ['notice', 'warning', 'error', 'critical', 'alert', 'emergency'] as const)
            expect(typeof la[k]).toBe('function');

        la.logger = logger2;
        expect(la.logger).toBe(logger2);

        la.severity = Severity.Warning;
        expect(la.severity).toBe(Severity.Warning);
        for (const k of ['debug', 'info', 'notice'] as const) expect(la[k]).toBe(undefined);
        for (const k of ['warning', 'error', 'critical', 'alert', 'emergency'] as const)
            expect(typeof la[k]).toBe('function');

        la.severity = null;
        expect(la.severity).toBe(null);
        for (const k of [
            'debug',
            'info',
            'notice',
            'warning',
            'error',
            'critical',
            'alert',
            'emergency',
        ] as const)
            expect(la[k]).toBe(undefined);

        la.severity = Severity.Debug;
        expect(la.severity).toBe(Severity.Debug);
        for (const k of [
            'debug',
            'info',
            'notice',
            'warning',
            'error',
            'critical',
            'alert',
            'emergency',
        ] as const)
            expect(typeof la[k]).toBe('function');

        la.logger = undefined;
        for (const k of [
            'debug',
            'info',
            'notice',
            'warning',
            'error',
            'critical',
            'alert',
            'emergency',
            'logger',
        ] as const)
            expect(la[k]).toBe(undefined);
        expect(la.severity).toBe(null);

        for (const {
            log: {
                mock: { calls },
            },
        } of [logger1, logger2]) {
            replaceTimedeltaFromCalls(calls);
            expect(calls).toMatchSnapshot();
        }
    });
});
