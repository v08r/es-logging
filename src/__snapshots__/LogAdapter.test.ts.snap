// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`LogAdapter\` lifecycle 1`] = `
[
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      300998,
      "logger level has been decreased from Error to Notice",
      null,
    ],
  ],
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      400999,
      "logger has been detached",
      null,
    ],
  ],
]
`;

exports[`\`LogAdapter\` lifecycle 2`] = `
[
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      300999,
      "logger has been attached",
      null,
    ],
  ],
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      400998,
      "logger level has been increased from Notice to Warning",
      null,
    ],
  ],
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      400998,
      "logger level has been increased from Warning to Off",
      null,
    ],
  ],
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      300998,
      "logger level has been decreased from Off to Debug",
      null,
    ],
  ],
  [
    [
      null,
      42,
      "",
      "productID",
      1,
      400999,
      "logger has been detached",
      null,
    ],
  ],
]
`;
